crates/ide/src/completion/complete_postfix/format_like.rs

```diff
@@ -0,0 +1,325 @@
+pub(super) fn add_format_like_completions(
+    acc: &mut Completions,
+    ctx: &CompletionContext,
+    dot_receiver: &ast::Expr,
+    cap: SnippetCap,
+    receiver_text: &str,
+) {
+    if !is_string_literal(receiver_text) {
+        // It's not a string literal, do not parse input.
+        return;
+    }
+
+    let input = &receiver_text[1..receiver_text.len() - 1];
+
+    let mut parser = FormatStrParser::new(input);
+
+    if parser.parse().is_ok() {
+        for kind in PostfixKind::all_suggestions() {
+            let snippet = parser.into_suggestion(*kind);
+            let (label, detail) = kind.into_description();
+
+            postfix_snippet(ctx, cap, &dot_receiver, label, detail, &snippet).add_to(acc);
+        }
+    }
+}
+
+/// Checks whether provided item is a string literal.
+fn is_string_literal(item: &str) -> bool {
+    if item.len() < 2 {
+        return false;
+    }
+    item.starts_with("\"") && item.ends_with("\"")
+}
+
+/// Parser for a format-like string. It is more allowing in terms of string contents,
+/// as we expect variable placeholders to be filled with expressions.
+#[derive(Debug)]
+pub struct FormatStrParser {
+    input: String,
+    output: String,
+    extracted_expressions: Vec<String>,
+    state: State,
+    parsed: bool,
+}
+
+#[derive(Debug, Clone, Copy)]
+pub enum PostfixKind {
+    Format,
+    Panic,
+    Println,
+    LogDebug,
+    LogTrace,
+    LogInfo,
+    LogWarn,
+    LogError,
+}
+
+impl PostfixKind {
+    pub fn all_suggestions() -> &'static [PostfixKind] {
+        &[
+            Self::Format,
+            Self::Panic,
+            Self::Println,
+            Self::LogDebug,
+            Self::LogTrace,
+            Self::LogInfo,
+            Self::LogWarn,
+            Self::LogError,
+        ]
+    }
+
+    pub fn into_description(self) -> (&'static str, &'static str) {
+        match self {
+            Self::Format => ("fmt", "format!"),
+            Self::Panic => ("panic", "panic!"),
+            Self::Println => ("println", "println!"),
+            Self::LogDebug => ("logd", "log::debug!"),
+            Self::LogTrace => ("logt", "log::trace!"),
+            Self::LogInfo => ("logi", "log::info!"),
+            Self::LogWarn => ("logw", "log::warn!"),
+            Self::LogError => ("loge", "log::error!"),
+        }
+    }
+
+    pub fn into_macro_name(self) -> &'static str {
+        match self {
+            Self::Format => "format!",
+            Self::Panic => "panic!",
+            Self::Println => "println!",
+            Self::LogDebug => "log::debug!",
+            Self::LogTrace => "log::trace!",
+            Self::LogInfo => "log::info!",
+            Self::LogWarn => "log::warn!",
+            Self::LogError => "log::error!",
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+enum State {
+    NotExpr,
+    MaybeExpr,
+    Expr,
+    MaybeIncorrect,
+    FormatOpts,
+}
+
+impl FormatStrParser {
+    pub fn new(input: impl Into<String>) -> Self {
```

Comment:
```suggestion
    pub fn new(input: String) -> Self {
```

https://github.com/rust-analyzer/rust-analyzer/blob/master/docs/dev/style.md#avoid-monomorphization